### ISSUE TITLE
[FSDP] `use_orig_params=True`: allow non-uniform `requires_grad` during init

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#91167]
For `use_orig_params=True`, to improve the UX, we can allow non-uniform `requires_grad` across the original parameters passed to construct one `FlatParameter` and save their `requires_grad` information. As long as one original parameter `requires_grad`, the current solution is to have the entire `FlatParameter` require gradient; however, we can still avoid forwarding a gradient to any original parameter that does not require gradient. The performant solution is to put frozen parameters in their own `FlatParameter`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501

**Comment by harry-stark** [comment]
@awgu Could we know any timeline for this feature?? Thanks!

**Comment by awgu** [comment]
> @awgu Could we know any timeline for this feature?? Thanks!

Let me try to get around to this in the next week or two.

**Comment by rohan-varma** [comment]
Assigning as I'll be working on this after discussion with @awgu

**Comment by psinger** [comment]
Just checking in if there is an update on this issue?

I assume this issue is also related: https://github.com/pytorch/pytorch/issues/95805

**Comment by awgu** [comment]
In the current state of frozen parameter handling for FSDP:
- For `use_orig_params=False`, each FSDP module must have uniform frozen or non-frozen parameters (flattened into one flat parameter).
- For `use_orig_params=True`, each FSDP module _may_ have mixed frozen/non-frozen parameters (flattened into one flat parameter), but the gradient memory will be higher than expected due to the corresponding 'flat gradient' having zeros for the frozen parameters.

@psinger @harry-stark I was wondering how you guys feel about the requirement that each FSDP module should have uniform frozen or non-frozen parameters.
1. Do you guys want to mix frozen/non-frozen because it is _hard_ to write a wrapping policy that does not mix or because it is _impossible_ to write one for your model (e.g. `bias` is frozen while `weight` is non-frozen for a single `nn.Linear`)?
2. Are you guys intentionally willing to pay the extra gradient memory cost? Or, is the extra gradient memory cost also a major pain point?

**Comment by maxjeblick** [comment]
> 1. Do you guys want to mix frozen/non-frozen because it is hard to write a wrapping policy that does not mix or because it is impossible to write one for your model (e.g. bias is frozen while weight is non-frozen for a single nn.Linear)?

I think this mixing of frozen/unfrozen params is exactly what happens when using LORA layers ([here](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora.py#L626)).

> 2. Are you guys intentionally willing to pay the extra gradient memory cost? Or, is the extra gradient memory cost also a major pain point?

`use_orig_params=True` seems to introduce a large memory overhead when using LORA where backbone is frozen. I compared `EleutherAI/pythia-2.8b-deduped` LORA training with both normal ddp and FSDP with `use_orig_params=True`, it turns out that FSDP consumes more GPU memory than DDP.
